### PR TITLE
docs: Add contributing guidelines for go-tuf

### PR DESCRIPTION
**Pull request by asraa**
Include PR process, review process, and other communication details.

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

File: CONTRIBUTING.md
Comment (mnm678) [inline CONTRIBUTING.md:50]: One day might be a bit optimistic. While this will usually be the case maintainers might take a week off, etc and I don't want to set expectations too high.

File: CONTRIBUTING.md
Comment (mnm678) [inline CONTRIBUTING.md:54]: we can also have something here about ensuring compliance with the TUF spec and overall code quality

File: CONTRIBUTING.md
Comment (mnm678) [inline CONTRIBUTING.md:23]: This section is a great addition, especially for new contributers

**Review by mnm678 — COMMENTED** [review]
Thanks for putting this together! I left some comments inline

File: CONTRIBUTING.md
Comment (rdimitrov) [inline CONTRIBUTING.md:43]: Maybe it would be nice to mention the community meeting schedule as well 👍

File: CONTRIBUTING.md
Comment (trishankatdatadog) [reply CONTRIBUTING.md:50]: How does a 5 business days sound? Also, it's unlikely all the maintainers are off, and I think we can make exceptions for regional holidays.

File: CONTRIBUTING.md
Comment (znewman01) [inline CONTRIBUTING.md:14]: Maybe: `...or Slack (see [Communication](#Communication))`

File: CONTRIBUTING.md
Comment (znewman01) [inline CONTRIBUTING.md:35]: If we require this, probably worth adding the [DCO bot](https://probot.github.io/apps/dco/)

Comment (trishankatdatadog) [reply CONTRIBUTING.md:35]: I _hate_ that DCO bot. Can we instead use an app that checks whether a contributor has signed a DCO agreement? Something like what the Linux Foundation uses.

Comment (znewman01) [reply CONTRIBUTING.md:35]: :rofl: :rofl:  I'm not attached to any particular bot, just saying that if it's "required" that should be enforced :)

Though I can't find what you're talking about -- I checked a few LF repos and the ones I found all use the same bot ([ex.](https://github.com/LF-Engineering/kelpy/pull/4/checks)).

Comment (trishankatdatadog) [reply CONTRIBUTING.md:35]: Sorry, I meant a CLA bot, which many orgs have their own apps for, here's the LF one: https://github.com/apps/lf-cla-application

Comment (trishankatdatadog) [reply CONTRIBUTING.md:35]: TLDR: no to DCO, yes to CLA

Comment (joshuagl) [reply CONTRIBUTING.md:35]: Why the preference for CLA? 

As I understand it (I am not a lawyer) both are assertions that the submitter has the right to submit a change and grants the right to the receiver to use the submitted change. A CLA is a legal contract while the DCO is a more lightweight statement of the contributor asserting ownership. As a consequence of a CLA being a legal contract, they can be a barrier to contributors (requiring, i.e., a corporate legal representative to sign-off).

Comment (trishankatdatadog) [reply CONTRIBUTING.md:35]: Simply because DCO is extremely annoying (must remember to `--sign-off`, otherwise rebase and force-push hell; does not work with commits done from the website), whereas CLA is a one-time deal.

**Comment by rdimitrov** [comment]
In regards to #234, we should add content in the guideline for these topics too:

**1. Convention for writing commit messages**

- Upon release, the changelog is built automatically and it will group commits based on a given regex and title.
- This is especially important when we are introducing breaking changes because such will be highlighted in e separate group on top of the changelog for any new release. 
- The available groups will be - `Breaking changes`, `Features`, `Bug fixes`, and `Others` 
- To mark that a commit should be part of a given group, its commit message should have some of the following keywords - `breaking`, `feat`, or `fix`. Everything else is added to the Others group.
- The responsibility for doing that is first to the author who created these commits and second to the maintainers reviewing and thus merging them. For example, if it introduces a breaking change and the commit messages do not inform of that, the maintainer should alter the message of the merge commit upon merging it so it includes the expected keyword.
 
**2. How to tag releases and prereleases**
- Use Semver-like versioning, i.e `v1.2.3` and `v1.2.3-rc1`?

**3. How to tag breaking changes**
- Increment the MAJOR version or something else?

@asraa - Let me know in case you are not available to spend some time on that, so we can merge this one with what we have so far and I'll address these in another PR 👍

**Comment by rdimitrov** [comment]
In regards to #234, we should add content in the guideline for these topics too:

**1. Convention for writing commit messages**

- Upon release, the changelog is built automatically and it will group commits based on a given regex and title.
- This is especially important when we are introducing breaking changes because such will be highlighted in e separate group on top of the changelog for any new release. 
- The available groups will be - `Breaking changes`, `Features`, `Bug fixes`, and `Others` 
- To mark that a commit should be part of a given group, its commit message should have some of the following keywords - `breaking`, `feat`, or `fix`. Everything else is added to the Others group.
- The responsibility for doing that is first to the author who created these commits and second to the maintainers reviewing and thus merging them. For example, if it introduces a breaking change and the commit messages do not inform of that, the maintainer should alter the message of the merge commit upon merging it so it includes the expected keyword.
 
**2. How to tag releases and prereleases**
- Use Semver-like versioning, i.e `v1.2.3` and `v1.2.3-rc1`?

**3. How to tag breaking changes**
- Increment the MAJOR version or something else?

@asraa - Let me know in case you are not available to spend some time on that, so we can merge this one with what we have so far and I'll address these in another PR 👍

**Comment by rdimitrov** [comment]
Since #234 got merged, maybe we should proceed with finalising this one too, so the contributing guidelines are consistent with what the CI expects to verify 👍

**Review by rdimitrov — APPROVED** [review]
@znewman01 - Let's merge this as it's already a great start for having such a guideline 👍

We'll address some of the comments in follow-up PRs and I'll open a separate issue for discussing whether we want to use CLA or DCO.

**Comment by znewman01** [comment]
I'd prefer to remove the contentious stuff (e.g. DCO vs. CLA) before we merge even a first-pass, so nobody confuses it for official, set-in-stone policy.

If we pull that out, I'm happy to approve it.

**Comment by rdimitrov** [comment]
> I'd prefer to remove the contentious stuff (e.g. DCO vs. CLA) before we merge even a first-pass, so nobody confuses it for official, set-in-stone policy.
> 
> If we pull that out, I'm happy to approve it.

Got it 👍 

@asraa - Would you like to address this one or if you don't have the time maybe I can fork your branch and do this on top?

**Comment by znewman01** [comment]
Superceded by #309

**Pull request closed without merging**
